### PR TITLE
Make the Scheduler adjust the steps taken relative to the gradient accumulation steps

### PR DESCRIPTION
**Pull request by muellerzr**
# Let the `AcceleratedScheduler` handle gradient accumulation steppage

## What does this add?

This PR adjusts the logic in the `AcceleratedScheduler` to take into account gradient accumulation steps. 

## Who is it for?

Closes https://github.com/huggingface/accelerate/issues/1170
Closes https://github.com/huggingface/accelerate/issues/1160

## Why is it needed?

Currently a behavior does not exist that will automatically "cut" the LR scheduler of a user if they pass in `gradient_accumulation_steps`, so unless they are careful and adjust their LR scheduler beforehand, they're not actually stepping properly with the lr scheduler.

### User-facing:

None, explicitly

### Internal structure:

`AcceleratedScheduler`'s `step` function now will run `n*num_processes` where `n==gradient_accumulation_steps` to account for the difference. 

To test performance, I ran the equivalent training of `gradient_accumulation_steps==2` on a bs of 16, and a regular batch size of 32, with negligible performance differences (accuracy += 0.5%, F1 of 0.3). Likely due to batch norm layers

## Usage Example:

When building the `Accelerator`, pass in `adjust_scheduler_to_accumulation` (default `False`) to enable this behavior:

```python
    accelerator = Accelerator(
        cpu=args.cpu, 
        mixed_precision=args.mixed_precision, 
        gradient_accumulation_steps=2,
        adjust_scheduler_to_accumulation=True,
    )
```

### CODE REVIEW

File: docs/source/usage_guides/gradient_accumulation.mdx [hunk 1]
```diff
@@ -111,6 +111,11 @@ You can remove all the special checks for the step number and the loss adjustmen
 
 As you can see the [`Accelerator`] is able to keep track of the batch number you are on and it will automatically know whether to step through the prepared optimizer and how to adjust the loss. 
 
+<Tip>
+  Typically with gradient accumulation, you would need to adjust the number of steps to reflect the change in total batches you are 
+  training on. 🤗 Accelerate will automatically do this for you, so long as you pass `adjust_scheduler_to_accumulation` to the [`Accelerator`] object's `__init__`.
```
Comment (sgugger) [inline docs/source/usage_guides/gradient_accumulation.mdx:116]:
```suggestion

Typically with gradient accumulation, you would need to adjust the number of steps to reflect the change in total batches you are 
training on. 🤗 Accelerate will automatically do this for you, so long as you pass `adjust_scheduler_to_accumulation` to the [`Accelerator`] object's `__init__`.

```

File: src/accelerate/accelerator.py [hunk 2]
```diff
@@ -143,6 +143,10 @@ class Accelerator:
             The number of steps that should pass before gradients are accumulated. A number > 1 should be combined with
             `Accelerator.accumulate`. If not passed, will default to the value in the environment variable
             `ACCELERATE_GRADIENT_ACCUMULATION_STEPS`.
+        adjust_scheduler_to_accumulation (`bool`, *optional*, defaults to `False`):
+            Whether or not the scheduler should be adjusted to the number of gradient accumulation steps. If `True`,
+            the scheduler's length will be reduced by `gradient_accumulation_steps`. If the scheduler already accounts
+            for the number of gradient accumulation steps, this should be set to `False`.
```
Comment (sgugger) [inline src/accelerate/accelerator.py:149]: We're starting to have too many arguments to `Accelerator`. This one should be documented waaaay below as it's not as important, and we should think of grouping it with the gradient_accumulation_steps in a `GradientAccumulationPlugin` maybe.

File: src/accelerate/scheduler.py [hunk 3]
```diff
@@ -42,10 +42,24 @@ class AcceleratedScheduler:
             Whether or not the dataloaders split one batch across the different processes (so batch size is the same
             regardless of the number of processes) or create batches on each process (so batch size is the original
             batch size multiplied by the number of processes).
+        adjust_scheduler_to_accumulation (`bool`, *optional*, defaults to `False`):
```
Comment (sgugger) [inline src/accelerate/scheduler.py:45]:
```suggestion
        adjust_to_accumulation (`bool`, *optional*, defaults to `False`):
```
We are in a scheduler.

**Review by sgugger — COMMENTED** [review]
Thanks for working on this. Just not sure about the API (adding yet another argument to the Accelerator) for this.

**Comment by muellerzr** [comment]
@sgugger good for review, a few notes:

- You'll notice that in the test for grad_sync changes, for it to pass (and for everywhere to pass in general), having accelerator use the new stepped scheduler to `True` is used, and on *every* script we have this behavior of what should exist is what we've advertised using. As such, I think it would be good to change the default to `True` not `False` here, as users are expecting one thing rather than the other
- Should we keep `gradient_accumulation_steps` as an arg in `Accelerator`? I'm leaning towards yes *if* we change the default to `True` for reasons above, though we can also just keep it there for usability. (In my eyes, users would likely not *need* to adjust the other param, hence we only need to expose `gradient_accumulation_steps` there

File: src/accelerate/utils/dataclasses.py [hunk 4]
```diff
@@ -380,6 +380,25 @@ def __post_init__(self):
...
 
+@dataclass
+class GradientAccumulationPlugin(KwargsHandler):
+    """
+    A plugin to configure gradient accumulation behavior.
+    """
+
+    num_steps: int = field(default=None, metadata={"help": "The number of steps to accumulate gradients for."})
+    adjust_scheduler: bool = field(
+        default=False,
```
Comment (sgugger) [inline src/accelerate/utils/dataclasses.py:391]: This should be True by default I think.

**Review by sgugger — APPROVED** [review]
Thanks for iterating! I agree that the default should be `True` for `adjust_scheduler`. The gradient accumulation API is quite new anyway, so it's okay if it's adjusted a bit like this.

**Review by sgugger — APPROVED** [review]
